src/JestProcessManagement/JestProcess.ts

```diff
@@ -82,9 +82,9 @@ export class JestProcess {
     this.stopRequested = true
     this.keepAliveCounter = 1
     this.jestSupportEvents.clear()
-    this.runner.closeProcess()
     return new Promise(resolve => {
       this.resolve = resolve
```

Comment:
additional note: the stop-process efficiency will become critical once we switch to the promise-base start/stop, thus we need to ensure there is no loose logic or inefficiency...

Reply:
If I'm reading the logic of `debuggerProcessExit` right, the `--this.keepAliveCounter` should decrement the previously-set `1` to `0` and skip the `if` that would try to restart the process (unless there's an edge case where `keepAliveCounter` is modified again somewhere between the `stop` call and `debuggerProcessExit` firing…?).

Reply:
I think you are right, but something seems convoluted...  the whole `keepAlive` and `keepAliveCounter` are unnecessary red herrings, which can probably be simplified:

1. don't think `keepAlive` needs to be an ivar, it was never used outside of the constructor.
1. inside the `startRunner()`, do we need both `exited` and `keepAliveCounter`?  when `keepAliveCounter` reaches 0, doesn't it mean exited?
1. does `keepAliveCounter` need to be an ivar? if we want to force-stop the process, isn't there a "stopRequested` ivar? It seems we won't need `keepAliveCounter` outside of the `startRunner()`, so the code would be a lot cleaner if we just pass it in like `startRunner(kepAliveCounter)`, a pretty standard self-captured count-down pattern. It should check `this.stopRequested` when deciding if to recursively calling itself with the decremented counter.

Reply:
I think `exited` exists to prevent the `onExitCallback` from being called more than once, on the edge case where `debuggerProcessExit` is called more than once? I agree though that the flow is not ideal and could be refactored with fewer variables. The `keepAliveCounter = 1` hack could be replaced with `this.stopRequested` having a higher priority in the listener than the `keepAlive` logic.

The `keepAlive` iVar seems to be used (read-only?) in `JestProcessManager`. It looks like @marcinczenko would know more about why it exists since he's the one who originally made it.

Reply:
When I was rewriting the process logic, I was also facing backward compatibility logic. If I remember correctly `keepAlive` is one of them. `startJestProcess` is always called with `keepAlive` being set to `true`, but for some reason at the time of creating it, we decided to have it as an option. Perhaps it can be factored out, which may remove some backward compatibility noise. As you can see, it is used only in the constructor of `JestProcess` to force `keepAliveLimit` to be `1` in case `keepAlive` argument is set to `false`.

Reply:
The promise in `JestProcess.stop` has been added at the very end of writing the process logic. Its intention was to support restarting Jest after updating the snapshots:

```javascript
  private detectedSnapshotErrors() {
    if (!this.pluginSettings.enableSnapshotUpdateMessages) {
      return
    }
    vscode.window
      .showInformationMessage('Would you like to update your Snapshots?', { title: 'Replace them' })
      .then(response => {
        // No response == cancel
        if (response) {
          this.jestProcess.runJestWithUpdateForSnapshots(() => {
            if (this.pluginSettings.restartJestOnSnapshotUpdate) {
              this.jestProcessManager.stopJestProcess(this.jestProcess).then(() => {
                this.startProcess()
              })
              vscode.window.showInformationMessage('Updated Snapshots and restarted Jest.')
            } else {
              vscode.window.showInformationMessage('Updated Snapshots. It will show in your next test run.')
            }
          })
        }
      })
  }
```

The promise logic was not added to `stopAll` because `stopAll` was not at that time intended to be used in the *restart* logic and more as "clean after yourself and stop forever" thing. The restart logic should be used in the same way as in the code snippet shown above.

I do not know the true intention of the:

```javascript
this.stopProcess()
setTimeout(() => {
  this.startProcess()
}, 500)
```

sequence, I suspect, that what I just said was not obvious and a higher level API has been used to achieve that. And since the promise logic is not present in `stopAll` hence the timeout.

Making `stopAll` waiting for all promises is in my opinion worth considering as this would make API more consistent. I am not aware of any hanging and I think the timeout was not used for such reason.

Reply:
thanks @marcinczenko for good background info update. 

from the code snippet:
```
             this.jestProcessManager.stopJestProcess(this.jestProcess).then(() => {
                this.startProcess()
              })
```
I wonder if you experienced issues when `startProcess()` is called without waiting for `stopJestProcess()`? I guess I am curious the underlying issue the promised-based stop() is trying to resolve, giving the jestProcess is not reused in the subsequent start()...

Reply:
I do not recall having any specific issues. I remember that there was some conversations about making all stuff more and more synchronous, so this was maybe the first step towards this. I remember I first implemented restarting without promise logic - it was added later and must have been influenced by the reviews. As well pointed at that times, leaning towards more synchronous behaviour lowers complexity and that was I think the main trigger. It may also have been that in this specific case, keeping the UI updates more synchronous is nicer user experience (but I am no longer sure if this was a factor).

Reply:
yes, I vaguely remember the conversation about making the jest process more sequential with Promise. Ok, so that confirmed we didn't create a delay for any specific issue but just following the best practice.